src/nvim/api/win_config.c

```diff
@@ -235,18 +315,118 @@ void nvim_win_set_config(Window window, Dict(float_config) *config, Error *err)
+  WinConfig fconfig = win->w_config;
+
+  bool is_split = (!HAS_KEY(config, float_config, relative) || striequal(config->relative.data, ""))
+                  && ((!HAS_KEY(config, float_config,
+                                external) && !fconfig.external) || !config->external)
+                  && (has_split || has_vertical || was_normal);
 
-  if (!parse_float_config(config, &fconfig, !new_float, false, err)) {
+  if (!parse_float_config(config, &fconfig, !was_normal || is_split, false, err)) {
     return;
   }
-  if (new_float) {
+  if (was_normal && !is_split) {
     if (!win_new_float(win, false, fconfig, err)) {
       return;
     }
     redraw_later(win, UPD_NOT_VALID);
+  } else if (is_split) {
+    win_T *parent = NULL;
+    if (!HAS_KEY(config, float_config, win) || config->win != -1) {
+      parent = find_window_by_handle(fconfig.window, err);
+      if (!parent) {
+        return;
+      } else if (parent->w_floating) {
+        api_set_error(err, kErrorTypeException, "Cannot split a floating window");
+        return;
+      }
+    }
+
+    WinSplit old_split = win_split_dir(win);
+    if (has_vertical && !has_split) {
+      if (fconfig.vertical) {
+        if (old_split == kWinSplitRight || p_spr) {
+          fconfig.split = kWinSplitRight;
+        } else {
+          fconfig.split = kWinSplitLeft;
+        }
+      } else {
+        if (old_split == kWinSplitBelow || p_sb) {
+          fconfig.split = kWinSplitBelow;
+        } else {
+          fconfig.split = kWinSplitAbove;
+        }
+      }
+    }
+    win->w_config = fconfig;
+
+    // If there's no vertical or split set, or if the split is the same as the old split,
+    // then we can just change the size of the window.
+    if ((!has_vertical && !has_split)
+        || (was_normal && !HAS_KEY(config, float_config, win) && old_split == fconfig.split)) {
+      if (HAS_KEY(config, float_config, width)) {
+        win_setwidth_win(fconfig.width, win);
+      }
+      if (HAS_KEY(config, float_config, height)) {
+        win_setheight_win(fconfig.height, win);
+      }
+      redraw_later(win, UPD_NOT_VALID);
+      return;
+    }
+
+    if (was_normal) {
+      // If the window is the last in the tabpage or `fconfig.win` is
+      // a handle to itself, we can't split it.
+      if (win->w_frame->fr_parent == NULL || (parent != NULL && parent->handle == win->handle)) {
+        api_set_error(err, kErrorTypeValidation, "Cannot split window into itself");
+        return;
+      }
+      int dir;
+      winframe_remove(win, &dir, NULL);
+      win_remove(win, NULL);
+    } else {
+      // Remove the old window from the tree of frames
+      win_remove(win, NULL);
+      // FIXME: Do I need to call this here? Seems to be the only way to close the float without
+      // freeing the window structure.
+      ui_comp_remove_grid(&win->w_grid_alloc);
+      assert(first_tabpage != NULL);  // suppress clang "Dereference of NULL pointer"
```

Comment:
```suggestion
```

Reply:
I am overall unsure of the chunk below `ui_comp_remove_grid` - it is copy-pasted from `win_close`, but if it looks good I'll just remove that line :)